fission/src/systems/scene/GizmoSceneObject.ts

```diff
@@ -0,0 +1,194 @@
+
+    private _size: number
+
+    /** @returns the instance of the transform gizmo itself */
+    public get gizmo() {
+        return this._gizmo
+    }
+
+    /** @returns Object3D that is attached to transform gizmo */
+    public get obj() {
+        return this._obj
+    }
+
+    /** @returns true if gizmo is currently being dragged */
+    public get isDragging() {
+        return this._gizmo.dragging
+    }
+
+    /** @returns the id of the parent scene object */
+    public get parentObjectId() {
+        return this._parentObject?.id
+    }
+
+    public constructor(
+        obj: THREE.Mesh = new THREE.Mesh(
+            new THREE.SphereGeometry(0.2),
+            new THREE.MeshBasicMaterial({ color: 0x00ff00 })
+        ),
+        mode: GizmoMode,
+        size: number,
+        parentObject?: MirabufSceneObject
+    ) {
+        super()
+
+        this._obj = obj
+        this._parentObject = parentObject
+        this._mainCamera = World.SceneRenderer.mainCamera
+
+        this._size = size
+
+        this._gizmo = new TransformControls(World.SceneRenderer.mainCamera, World.SceneRenderer.renderer.domElement)
+        this._gizmo.setMode(mode)
+
+        World.SceneRenderer.RegisterGizmoSceneObject(this)
+    }
+
+    public Setup(): void {
+        // adding the mesh and gizmo to the scene
+        World.SceneRenderer.AddObject(this._obj)
+        World.SceneRenderer.AddObject(this._gizmo)
+
+        // forcing the gizmo to rotate and transform with the object
+        this._gizmo.setSpace("local")
+        this._gizmo.attach(this._obj)
+
+        this._gizmo.addEventListener("dragging-changed", (event: { target: TransformControls; value: unknown }) => {
+            // disable orbit controls when dragging the transform gizmo
+            const gizmoDragging = World.SceneRenderer.IsAnyGizmoDragging()
+            if (!event.value && !gizmoDragging) {
+                World.SceneRenderer.orbitControls.enabled = true // enable orbit controls when not dragging another transform gizmo
+            } else if (!event.value && gizmoDragging) {
+                World.SceneRenderer.orbitControls.enabled = false // disable orbit controls when dragging another transform gizmo
+            } else {
+                World.SceneRenderer.orbitControls.enabled = !event.value // disable orbit controls when dragging transform gizmo
+            }
+
+            if (event.target.mode === "translate") {
+                // disable other gizmos when translating
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (obj.gizmo.object === event.target.object && obj.gizmo.mode !== "translate") {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
```

Comment:
Since you have the `SceneRenderer._gizmosOnMirabuf` map, you should no longer filter `sceneObjects`. In this case, you should make it so that you either make `_gizmosOnMirabuf` publicly available or have a method in SceneRenderer to test if some object has a gizmo (which would be as simple as `this._gizmosOnMirabuf.has(mirabuf_id)` from inside SceneRenderer).

As long as the mirabuf instance id that would be the key in `gizmosOnMirabuf` is available inside this function that I'm commenting on, you can then reduce this from O(n) to O(probably 1) since it would only require a Map lookup instead of an array filter and then foreach.